File: frontend_mobile/lib/data/models/log_post/log_post_detail_response_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pairing_planet2_frontend/data/models/image/image_response_dto.dart';
import 'package:pairing_planet2_frontend/data/models/recipe/recipe_summary_dto.dart';
import 'package:pairing_planet2_frontend/data/models/hashtag/hashtag_dto.dart';
import 'package:pairing_planet2_frontend/domain/entities/log_post/log_post_detail.dart';

part 'log_post_detail_response_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class LogPostDetailResponseDto {
  final String publicId;
  final String? title;
  final String content;
  final String? outcome; // SUCCESS, PARTIAL, FAILED (nullable for backward compat)
  final List<ImageResponseDto>? images;
  final RecipeSummaryDto? linkedRecipe;
  final String createdAt;
  final List<HashtagDto>? hashtags;
  final bool? isSavedByCurrentUser;
  final String? creatorPublicId; // For ownership check (UUID string)
  final String? userName;     // Creator's username for display

  LogPostDetailResponseDto({
    required this.publicId,
    required this.title,
    required this.content,
    this.outcome,
    required this.images,
    required this.linkedRecipe,
    required this.createdAt,
    this.hashtags,
    this.isSavedByCurrentUser,
    this.creatorPublicId,
    this.userName,
  });

  factory LogPostDetailResponseDto.fromJson(Map<String, dynamic> json) =>
      _$LogPostDetailResponseDtoFromJson(json);
  Map<String, dynamic> toJson() => _$LogPostDetailResponseDtoToJson(this);

  LogPostDetail toEntity() {
    return LogPostDetail(
      publicId: publicId,
      content: content,
      outcome: outcome ?? 'PARTIAL', // Default to PARTIAL if null
      imageUrls: images?.map((img) => img.imageUrl).toList() ?? [],
      imagePublicIds: images?.map((img) => img.imagePublicId).toList() ?? [],
      recipePublicId: linkedRecipe?.publicId ?? "",
      linkedRecipe: linkedRecipe != null
          ? LinkedRecipeInfo.fromRecipeSummary(linkedRecipe!.toEntity())
          : null,
      createdAt: DateTime.parse(createdAt),
      hashtags: hashtags?.map((e) => e.toEntity()).toList() ?? [],
      isSavedByCurrentUser: isSavedByCurrentUser,
      creatorPublicId: creatorPublicId,
      userName: userName,
    );
  }
}

File: backend/src/main/java/com/cookstemma/cookstemma/dto/log_post/LogPostDetailResponseDto.java
package com.cookstemma.cookstemma.dto.log_post;

import com.cookstemma.cookstemma.dto.hashtag.HashtagDto;
import com.cookstemma.cookstemma.dto.image.ImageResponseDto;
import com.cookstemma.cookstemma.dto.recipe.RecipeSummaryDto;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * Log post detail DTO with pre-localized title/content.
 * The title and content fields contain values for the requested locale,
 * resolved server-side from the translations maps.
 */
public record LogPostDetailResponseDto(
        UUID publicId,
        String title,        // Localized title
        String content,      // Localized content
        Integer rating,      // 1-5 star rating
        List<ImageResponseDto> images,
        RecipeSummaryDto linkedRecipe,
        Instant createdAt,
        List<HashtagDto> hashtags,
        Boolean isSavedByCurrentUser,  // null if not logged in
        UUID creatorPublicId,          // for ownership check (UUID for frontend comparison)
        String userName,               // Creator's username for display
        Boolean isPrivate              // Whether this log is private
) {}